app/services/embedded_page_service.rb

```diff
@@ -0,0 +1,90 @@
+# Processes requests for pages embedded in iframes
+
+class EmbeddedPageService
+  attr_reader :use_embedded_layout
+
+  def initialize(params, session, request, response)
+    @params = params
+    @session = session
+    @request = request
+    @response = response
+
+    @embedding_domain = @session[:embedding_domain]
+    @use_embedded_layout = false
+  end
+
+  def embed!
+    return unless embeddable?
+    return if embedding_without_https?
+
+    process_embedded_request
+    set_response_headers
+    set_embedded_layout
+  end
+
+  private
+
+  def embeddable?
+    return true if current_referer == @request.host
```

Comment:
things are more maintainable and easier to change in the future if we use `attr_reader` for the ivars.

Reply:
I'm not sure what you want in `attr_reader`? 

The request object is passed in to the class and altered in various places, but doesn't need to be explicitly read from outside of the class. It needs to be available to the various methods internally but it doesn't need to be readable...

Reply:
My point is to use private reader methods instead of the raw ivars. That brings the usual benefits of encapsulation.

Reply:
I'm not sure I understand what you mean? Adding `attr_reader` and `attr_accessor` lines under `private` will just create private methods for accessing those instance variables, right? But they can't be accessed outside of the instance anyway, and all but one of the methods are private to begin with. 

The class needs to be able to modify the objects referenced by `@resonse` and `@session`. Actually, that's basically it's primary job. It looks like it only needs to read the other two objects, so I guess we could use:

```
private
attr_reader :request, :params
attr_accessor :session, :response
```
But I don't really understand what would be gained from this? There's already no getters or setters for those variables, so this doesn't limits access in any way. Is the intention to make it more explicit that those variables are only accessible from within an instance of this class?

Reply:
I think the intention here is to standardise access to always use methods instead of direct access to the ivars. The advantage is that it's easier to change behaviour later or move parts into other classes without changing the code accessing these variables. The disadvantage is that you can't see immediately if a call to `request` has side effects or if it is private. You need to look up the accessor/reader first. That's why I'm not clear on this recommendation.

Reply:
Yes, that's my point. 

Regarding

> The disadvantage is that you can't see immediately if a call to request has side effects or if it is private. You need to look up the accessor/reader first. 

I haven't experienced problems with this unless the class is huge. Once you open the class, from the constructor you see what vars compose its internal state and what is all about. However, this encapsulation brings much bigger benefits to changeability. 

It's relatively easier to move private methods to the public API without having to change much, or create a new subclass and rely on those reader methods or even to turn simple `attr_reader`s to regular methods when their logic grows, while keeping the consumers untouched. Not to mention the fact that a typo on a method call raises an error, while a typo on an ivar's name does not; that's really useful!

Having said that, I don't want to be picky. You guys decide.

Reply:
It looks like I would also have to convert lines like this:
`@embedding_domain = current_referer`

to use this syntax:
`instance_variable_set '@embedding_domain', current_referer`

because this won't work for assigning to the private ivars via methods:
`embedding_domain = current_referer`

Reply:
I'm not sure I know what you mean. You last line doesn't work even if you have a private `attr_writer`? Anyway, I'm moving this to test ready and keep all the comments in mind. I will probably make more use of accessor in the future.

Reply:
I don't follow it either. Look at https://github.com/codetriage/codetriage/blob/master/app/services/git_hub_authenticator.rb. This is a clear example of what I mean.

Reply:
It works nicely when reading from ivars via private attr_reader methods, but writing to those attributes needs some odd-looking rewrites using `#instance_variable_set`:
```ruby
private
attr_reader :private_ivar_one
attr_accessor :private_ivar_two

def example
example_local_variable = private_ivar_one # reading works fine and is clear/concise

private_ivar_two = 'new_value' # doesn't write to ivar, creates new local variable
instance_variable_set @private_ivar_two, 'new_value' # writes to the ivar, but isn't pretty
end
```

Reply:
Now I understand what you mean. That's what `self` is for. Check out [this](http://rubyfiddle.com/riddles/34ec5) example.

Reply:
Ah that's really obvious now, I don't know why I was only thinking of more complicated solutions like `send(private_ivar='new_value')` and everything except the simple answer... :see_no_evil: